spec/dog_spec.rb

```diff
@@ -0,0 +1,102 @@
+require "spec_helper"
+require 'pry'
+
+describe "Dog" do
+
+  let(:teddy) {Dog.new("Teddy", "cockapoo")}
+
+  before(:each) do
+    DB[:conn].execute("DROP TABLE IF EXISTS dogs")
+    sql =  <<-SQL
+      CREATE TABLE IF NOT EXISTS dogs (
+        id INTEGER PRIMARY KEY,
+        name TEXT,
+        breed TEXT
+        )
+    SQL
+    DB[:conn].execute(sql)
+  end
+
+  describe "attributes" do
+    it 'has a name and a breed' do
+      dog = Dog.new("Fido", "lab")
+      expect(dog.name).to eq("Fido")
+      expect(dog.breed).to eq("lab")
+    end
+
+    it 'has an id that defaults to `nil` on initialization' do
+      expect(teddy.id).to eq(nil)
+    end
+  end
+
+  describe "#create_table" do
+    it 'creates the dogs table in the database' do
+      Dog.create_table
+      table_check_sql = "SELECT tbl_name FROM sqlite_master WHERE type='table' AND tbl_name='dogs';"
+      expect(DB[:conn].execute(table_check_sql)[0]).to eq(['dogs'])
+    end
+  end
+
+  describe "#drop_table" do
+    it 'drops the dogs table from the database' do
+      Dog.drop_table
+      table_check_sql = "SELECT tbl_name FROM sqlite_master WHERE type='table' AND tbl_name='dogs';"
+      expect(DB[:conn].execute(table_check_sql)[0]).to eq(nil)
+    end
+  end
+
+  describe "#save" do
+    it 'saves an instance of the dog class to the database and then sets the given dogs `id` attribute' do
+      teddy.save
+      expect(DB[:conn].execute("SELECT * FROM dogs")).to eq([[1, "Teddy", "cockapoo"]])
+      expect(teddy.id).to eq(1)
+    end
+  end
+
+  describe "#create" do
+    it 'takes in a hash of attributes and uses metaprogramming to create a new dog object. Then it uses the #save method to save that dog to the database' do
+      Dog.create(name: "Ralph", breed: "lab")
+      expect(DB[:conn].execute("SELECT * FROM dogs")).to eq([[1, "Ralph", "lab"]])
+    end
+  end
+
+  describe '.find_or_create_by' do
+    it 'creates an instance of a dog if it does not already exist' do
+      dog1 = Dog.create(name: 'teddy', breed: 'cockapoo')
+
+      dog2 = Dog.find_or_create_by(name: 'teddy', breed: 'cockapoo')
+      expect(dog1.id).to eq(dog2.id)
```

Comment:
handful more tests needed here:

  - [ ] Creates one if it doesn't exist
  - [ ] Add in two dogs with the same name, different breed, then do a `find_or_create_by` on the name and one of the breeds, make sure the id is correct and it returns the right one!
  - [ ] Add in two dogs with the same name, different breed, then `find_or_create_by` another dog with the same name and a different breed, make sure it's a new one